### ISSUE TITLE
Running tests with  an interactive terminal (noninteractive: nil)

### ISSUE THREAD

**Issue by cpitclaudel** [jorgenschaefer/emacs-buttercup#174]
I have tests that can only run when `noninteractive` is `nil` (because they use `format-mode-line`, which just returns `""` in noninteractive mode).  Is there a way to run these with buttercup?

**Comment by snogge** [comment]
Try binding `noninteractive` to true in your test.

**Comment by cpitclaudel** [comment]
That won't do anything, will it?

**Comment by cpitclaudel** [comment]
```
$ emacs -Q --batch --eval '(print (format-mode-line (quote "A")))'

""
$ emacs -Q --batch --eval '(print (let ((noninteractive nil)) (format-mode-line (quote "A"))))'

""
```

(you meant `noninteractive` to `nil`, not `t`, I think)

**Comment by snogge** [comment]
I did mean nil.
I think it does not work because `format-mode-line` is a C function.  I don't think there is any way around this as long as you run Emacs in batch mode.

**Comment by cpitclaudel** [comment]
>  I don't think there is any way around this as long as you run Emacs in batch mode.

I think that's right; I was hoping for a way to tell the buttercup binary to start in non-batch mode (e.g. using an invisible frame).

**Comment by snogge** [comment]
The buttercup binary (bin/buttercup) is just a bash script.  
You can start a daemonized emacs (which has a hidden frame if I remember correctly) and execute elisp on that daemonized instance:

```bash
emacs -Q --daemon=foo
emacs -Q --batch --eval "(message \"%s\" (progn (require 'server) (server-eval-at \"foo\" '(format-mode-line (quote \"A\")))))"
emacs -Q --batch --eval "(progn (require 'server) (server-eval-at \"foo\" '(kill-emacs)))"
```

I have hacked together a way to actually run some simple buttercup tests on the Emacs daemon, but so far have not even started on sending the messages back to the batch Emacs for printing on the terminal.

**Comment by snogge** [comment]
But for your tests it should be possible to start an Emacs daemon before you run `buttercup`and then let the individual tests use `server-eval-at`.

**Comment by snogge** [comment]
@cpitclaudel, is it OK to close this issue?   Or do you think
something should be changed in or added to buttercup.el?

**Comment by cpitclaudel** [comment]
Ideally, buttercup would do the server-eval-at itself, but wishing is easy :)  We could close this.

**Comment by codygman** [comment]
If you want to do this, I run emacs with `emacs -nw` and use this function:

```elisp
(defun tests-run ()
  ;; We would like to use `ert-run-tests-batch-and-exit'
  ;; Unfortunately it doesn't work outside of batch mode, and we
  ;; can't use batch mode because we have tests that need windows.
  ;; Instead, run the tests interactively, copy the results to a
  ;; text file, and then exit with an appropriate code.
  (setq attempt-stack-overflow-recovery nil
	attempt-orderly-shutdown-on-fatal-signal nil
	debug nil
	debug-on-error t
	)
  (unwind-protect
      (progn
	(ert-run-tests-interactively t)
	(with-current-buffer "*ert*"
	  (append-to-file (point-min) (point-max) "test-results.txt")
	  (let ((failed-tests (ert-stats-completed-unexpected ert--results-stats)))
	    ;; (log (format "failed tests: %s" (princ failed-tests)))
	    ;; (log (format "zerop failed-tests ==  %s" (princ (zerop failed-tests))))
	    (log "checking failed tests")
	    (if (zerop failed-tests)
		(progn (log "0 failed tests, exiting with code 0")
		       (when (not debug) (kill-emacs 0)))
	      (progn (log "at least 1 failed test, exiting with code 1")
		     (when (not debug) (kill-emacs 1))))
	    (log "done checking failed tests")
```

I soon want to try and get it working with buttercup, but in the short term I might just split out my tests that I can't run with buttercup in batch mode with the function above.

It would be nice if buttercup supported this mode by default for testing evil functions or the one you needed I think.

**Comment by codygman** [comment]
Do you have your daemon work up anywhere?

My idea was just to use `async.el`, but I guess it might also startup an emacs without interactive support. That would mean running an emacs daemon which I'm gathering is like `emacs -nw` and can support tests like this.

I'm getting very annoyed (as in enough to try helping add this feature to buttercup) with not being able to use buttercup for most of my useful tests.

One example of a useful test I can't run with buttercup:

```
(it "flycheck squiggly appears underneath misspelled putStrLnORAORAORA function"
     (find-file (emacs-d-directory-for "testdata/simple-haskell-project/Main.hs"))
     (replace-string "putStrLn" "putStrLnORAORAORA")
     (save-buffer)
     (sit-for 1)
     (expect (get-char-property (point) 'face) :to-equal 'flycheck-error))
```

results in:

```shell
Codygman's hci Haskell Integration flycheck squiggly appears underneath misspelled putStrLnORAORAORA function

Traceback (most recent call last):
  (perform-replace "putStrLn" "putStrLnORAORAORA" nil nil nil nil nil nil ni...
  (replace-match-maybe-edit "putStrLnORAORAORA" t t nil (69 77 #<buffer Main...
  (replace-match "putStrLnORAORAORA" t t)
  (ask-user-about-lock "/home/cody/hci/testdata/simple-haskell-project/Main....
  (error "Cannot resolve lock conflict in batch mode")
  (signal error ("Cannot resolve lock conflict in batch mode"))
error: (error "Cannot resolve lock conflict in batch mode")
```

**Comment by codygman** [comment]
Maybe `emacs -nw` and capabilities of `emacs --daemon` aren't equivalent for this specific flycheck case. I tried out the solution of starting an emacs daemon. See the [commit](https://github.com/codygman/hci/commit/e72b376715dcd4f3f1c2c556ca3a213439cb9a8a)/[CI](https://github.com/codygman/hci/runs/837735883?check_suite_focus=true).

**Comment by snogge** [comment]
Sorry, I can't find the daemon code at the moment.  Perhaps I deleted
it by accident, which would be a shame. 

Following the commit I think you are going about the test the wrong
way.  Do you really need to verify the whole chain of "edit, wait for
flycheck to react, check that the highlighting is correct" in a single
test?  You are basically redoing the testing of flycheck.
Or maybe I just totally misunderstand what you are trying to do and/or
the purpose of this specific test.

Is this 'just' about testing a flycheck checker?  Because I'm sure
there are some patterns in the flycheck repo that you can use.
flycheck uses buttercup for at least part of their testing.

**Comment by codygman** [comment]
>  Do you really need to verify the whole chain of "edit, wait for
flycheck to react, check that the highlighting is correct" in a single
test? 

Yes. It's an integration test of flycheck with my increasingly complex config.

> You are basically redoing the testing of flycheck.

It's easy to think that, but I've had numerous issues with my own config that caused flycheck to stop working. The most recent being when I turned on a flycheck setting with lsp-haskell 1.9, flycheck stopped automatically highlighting correctly and the highlighting only showed up after manually disabling/re-enabling flycheck.

I'm using buttercup where I can to write integration tests from the highest level for the most crucial parts of my emacs config where I've had issues before. At one point I had a very nice config that I depended heavily on, but for some reason things would just break at the worst moments possible and I begun to not trust my system.

From the outside though, I understand your perspective and how you could see those tests or using nix to ensure everything is totally reproducible along with very granular git commits and those integration tests to very quickly spot problem.

**Comment by codygman** [comment]
Since it may be of interest, it looks like for some reason (perhaps the same as me) flycheck uses ert for these types of tests. I did find a much better way of waiting on flycheck in their tests, [flycheck-ert-wait-for-syntax-checker](https://github.com/flycheck/flycheck/blob/e7b0b704c97e8903bce2ea739c3e28979d376969/flycheck-ert.el#L298).

**Comment by lastquestion** [comment]
IMHO, `buttercup` is great for unit tests. Integration tests or any tests that require testing the full stack should not exist in buttercup, but in some "other testing framework".

Buttercup == mocha
??? = Selenium / Cypress

For me, I implemented a mini test framework for ??? . I run a subprocess emacs inside tmux frames for full integration tests.
It's super powerful because you can send literal key presses, verify anything you like in actual display output by asking tmux take a screenshot if you need, etc. Plus, it's literally identical to what a user would do.

Running `-nw` just means run without windowing. In that case, you already have different behavior than windowed emacs, but normally it's "not too much different" (except for the rare cases when it is...)

Running `daemon` and attaching a new client is definitely not identical. Running connected to a daemon means `server.el` runs, which may not act like you expect. In particular, how `server-eval` works can be quite brittle. It is not the same as actually executing an expression via `M-:` or `scratch`. Also, `-f` does not work like `init.el`, nor does `-f` work like `m-:` or calling `eval` directly.

**Comment by codygman** [comment]
@lastquestion Thank you for the info!

Any public code you can share for your tmux/emacs setup? I imagine verifying things like "functionnnn has red flycheck squiggly under it" could be annoying.

**Comment by lastquestion** [comment]
@codygman the driver code is https://github.com/lastquestion/explain-pause-mode/blob/master/tests/cases/driver.el, but it's not very documented and kind of still early days. Most of it is related to the package under test, the actual code that runs tmux is mostly here https://github.com/lastquestion/explain-pause-mode/blob/master/tests/cases/driver.el#L176 and there's code to send key presses and such further on. I reuse an existing mechanism that package exposes to communicate with the subprocess using UNIX sockets (which you could also take inspiration from), as well as using SIGUSR1.

Remember, tmux is `-nw`, so you don't get red squigglies. Two ways to test that: You can write a function that runs inside the emacs under test that reads under point, and checks the face property that it is squiggle. You can connect that to SIGUSR1 for example, and fire a signal to the emacs under test. This avoids `eval` and `m-x`, so this mostly is identical to a user experience __but not exactly__. For example, if you depend on idle timers to fire, etc. this might not work (I did not check if SIGUSR1 resets the idle timer.) 

Or you can ask tmux to take a screenshot and then check to see if the text you're checking for squiggles is underlined - which i think the default face for flycheck errors etc is just underline. I didn't do it in that `driver.el` yet because I am not sure if I need to but I did it previously for another project in another language. It's well documented though just `man tmux` and search for `capture-pane`.  

This whole discussion is probably a little off topic for this issue 😁 feel free to email me or open an issue on that project about exposing or packaging the test code out into another package or something.

**Comment by AutumnalAntlers** [comment]
https://github.com/AutumnalAntlers/emacs-buttercup/tree/my-fork

Based on the advice in this issue, I've been using buttercup to test functions that fundamentally depend on frames, posn-at-point, format-mode-line, etc. My script spawns an emacs server, uses emacsclient to run specs, and handles control flow through a named pipe. The biggest issue I encountered was that `emacsclient --eval foobar` doesn't invoke the debugger, so I had to modify `beacon--debugger` to catch, filter, and sometimes handle signals before they're caught by other handlers when tests are failing. I have tests that use `:to-throw`, so I know those work correctly with my patch, but backtraces do seem to go back a little further than they did without it.
